Add initial LWR AMR test problems.

nuclearkevin opened the pull request:
This PR contains the LWR AMR test cases I've been using up until this point to test various algorithms and to ensure that the AMR implementation is functioning as intended.

Closes #2, Closes #3.

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,149 @@
+                help='Whether to add a Shannon entropy mesh')
+args = ap.parse_args()
+
+N = args.n_axial # Number of axial cells to build in the solid to receive feedback
```
I'd prefer a variable name that makes it clear what this is: e.g. `num_axial`

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,149 @@
+args = ap.parse_args()
+
+N = args.n_axial # Number of axial cells to build in the solid to receive feedback
+height = 192.78  # Total height of pincell
+
+# Some geometric properties that can be modified to change the model.
+## The radius of a fuel pin (same for all pin types).
+r_fuel = 0.4095
+
+## The thickness of the fuel-clad gap.
+t_f_c_gap = 0.0085
+
+## The thickness of the Zr fuel pin cladding.
+t_zr_clad = 0.057
+
+## The thickness of the axial reflector above the lattice.
+reflector_t = 21.42
```
I thought there was a strategy where we could define these once in a file that is imported both to MOOSE (.i files) and to python?

lewisgross1296 replied:
The way I do it (stole this from the Cardinal tutorials) is to define a file `common_input.i` that can be included into a MOOSE input file via 
```
!include common_input.i
```
Then, you can make a symbolic link `ln -s common_input.i common_input.py` in the same directory (or elsewhere). That allows you to add a line like `import common_input as specs` at the top of your python scripts to grab the same list of parameters

nuclearkevin replied:
Thanks @lewisgross1296, this worked like a charm!

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,149 @@
+box = openmc.model.RectangularPrism(pitch, pitch, boundary_type='reflective')
+
+# Create cells, mapping materials to regions - split up the axial height
+planes = np.linspace(0.0, height, N + 1)
+plane_surfaces = []
+for i in range(N + 1):
+  plane_surfaces.append(openmc.ZPlane(z0=planes[i]))
```
```suggestion
plane_surfaces = [ openmc.ZPlane(z0=z) for z in np.linspace(0.0, height, N + 1) ]
```

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,149 @@
+  gap_cells.append(openmc.Cell(fill=None, region=+fuel_or & -clad_ir & layer, name='Gap{:n}'.format(i)))
+  clad_cells.append(openmc.Cell(fill=zr, region=+clad_ir & -clad_or & layer, name='Clad{:n}'.format(i)))
+  water_cells.append(openmc.Cell(fill=h2o, region=+clad_or & layer & -box, name='Water{:n}'.format(i)))
+  all_cells.append(fuel_cells[i])
+  all_cells.append(gap_cells[i])
+  all_cells.append(clad_cells[i])
+  all_cells.append(water_cells[i])
```
Is there a reason to append these in this order versus appending the full list of `fuel_`, `gap_`, `clad_` and `water_cells` after the loop?

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,149 @@
+gap_cells = []
+water_cells = []
+all_cells = []
+for i in range(N):
+  layer = +plane_surfaces[i] & -plane_surfaces[i + 1]
+  fuel_cells.append(openmc.Cell(fill=uo2, region=-fuel_or & layer, name='Fuel{:n}'.format(i)))
+  gap_cells.append(openmc.Cell(fill=None, region=+fuel_or & -clad_ir & layer, name='Gap{:n}'.format(i)))
+  clad_cells.append(openmc.Cell(fill=zr, region=+clad_ir & -clad_or & layer, name='Clad{:n}'.format(i)))
+  water_cells.append(openmc.Cell(fill=h2o, region=+clad_or & layer & -box, name='Water{:n}'.format(i)))
```
```suggestion
for layer_idx, lower_plane, upper_plane in enumerate(zip(plane_surfaces[:-1], plane_surfaces[1:])):
  layer = +lower_plane & - upper_plane
  fuel_cells.append(openmc.Cell(fill=uo2, region=-fuel_or & layer, name=f'Fuel{layer_idx}')
  gap_cells.append(openmc.Cell(fill=None, region=+fuel_or & -clad_ir & layer, name=f'Gap{layer_idx}')
  clad_cells.append(openmc.Cell(fill=zr, region=+clad_ir & -clad_or & layer, name=f'Clad{layer_idx}')
  water_cells.append(openmc.Cell(fill=h2o, region=+clad_or & layer & -box, name=f'Water{layer_idx}')
```

gonuke reviewed (changes requested):
Thanks for sharing these @nuclearkevin.  A lot things to look at, at once.

I've made a few suggestions to the single pin cell OpenMC model file, but they can perhaps be applied throughout?  It also seems like there may be some common dimensions and data that apply to all the LWR problems that could be defined just once and included/imported as needed.

nuclearkevin commented:
Thanks for the review @gonuke! I believe I've addressed all your comments (and propagated the changes out to the other LWR Python scripts). Feel free to drop another review if you have any other questions / concerns!

gonuke commented:
In my group, we'd definitely be discussing a lot more reuse of some of the basic building blocks of these files.  e.g. 
* the dimensions are borrowed across many of the LWR problems
* the materials are similar across many of the problems
* the definition of the basic pin cell is reused
* etc..

Is there any appetite for a review of that nature, consolidating information in fewer places and reusing/including/importing as appropriate/necessary?

nuclearkevin commented:
I'm definitely on-board with reducing code duplication as much as possible. I just pushed some changed to that effect - if you spot anything else that should be trimmed down please let me know and I'll jump on it!

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,64 @@
+import openmc_pincells as pins
+
+ap = ArgumentParser()
+ap.add_argument('-n', dest='n_axial', type=int, default=1,
```
Is it intentional that you allow this to be different from `AXIAL_DIVISIONS` defined in `common.i`?  Perhaps this is testing some of the Cardinal mapping between OpenMC and MOOSE?

nuclearkevin replied:
This was done intentionally to allow for the testing of multiphysics feedback with AMR. We'll need a relatively fine axial discretization of the CSG geometry to better capture temperature variations as they get projected from the adapted mesh.

gonuke commented on models/lwr/openmc_materials.py:
```diff
@@ -0,0 +1,109 @@
+#--------------------------------------------------------------------------------------------------------------------------#
+# Material definitions.
+## First fuel region: 4.3% MOX.
+mox_4_3_comp = 1.0e24 * np.array([5.00e-5, 2.21e-2, 1.50e-5, 5.80e-4, 2.40e-4, 9.80e-5, 5.40e-5, 1.30e-5, 4.63e-2])
+mox_4_3_frac = mox_4_3_comp / np.sum(mox_4_3_comp)
+mox_4_3 = openmc.Material(name = '4.3% MOX Fuel', temperature = 293.15)
+mox_4_3.add_nuclide('U235', mox_4_3_frac[0], percent_type = 'ao')
+mox_4_3.add_nuclide('U238', mox_4_3_frac[1], percent_type = 'ao')
+mox_4_3.add_nuclide('Pu238', mox_4_3_frac[2], percent_type = 'ao')
+mox_4_3.add_nuclide('Pu239', mox_4_3_frac[3], percent_type = 'ao')
+mox_4_3.add_nuclide('Pu240', mox_4_3_frac[4], percent_type = 'ao')
+mox_4_3.add_nuclide('Pu241', mox_4_3_frac[5], percent_type = 'ao')
+mox_4_3.add_nuclide('Pu242', mox_4_3_frac[6], percent_type = 'ao')
+mox_4_3.add_nuclide('Am241', mox_4_3_frac[7], percent_type = 'ao')
+mox_4_3.add_element('O', mox_4_3_frac[8], percent_type = 'ao')
+mox_4_3.set_density('atom/cm3', np.sum(mox_4_3_comp))
```
I worry about this being error prone since you rely on the indexing of the comp vector to line up with the order they are used.  Another alternative could be something like this (subject to final debugging):
```suggestion
mox_4_3_comp = {
  'U235': 5.00e-5,
  'U238': 2.21e-2,
  'Pu238': 1.50e-5,
  'Pu239': 5.80e-4,
  'Pu240': 2.40e-4,
  'Pu241': 9.80e-5,
  'Pu242': 5.40e-5,
  'Am241': 1.30e-5,
  'O': 4.63e-2
  }
mox_4_3_density = 1.0e24 * np.array(mox_4_3_comp.values())
mox_4_3 = openmc.Material(name = '4.3% MOX Fuel', temperature = 293.15)
for comp, atom_frac in mox_4_3_comp.items():
    if comp[-1].isdigit():
        mox_4_3.add_nuclide(comp, atom_frac, percent_type = 'ao')
    else:
        mox_4_3.add_element(comp, atom_frac, percent_type = 'ao')
mox_4_3.set_density('atom/cm3', mox_4_3_density)
```

This would also allow you to reuse the composition dictionary as a starting point to define the other MOX vectors.

The next step might be to define the composition dictionaries and the material names in a data structure that just lets you loop over them all to generate them.

nuclearkevin replied:
I've gone ahead and implemented this approach alongside a loop over all compositions to populate a dictionary of OpenMC materials.

gonuke commented on models/lwr/openmc_pincells.py:
```diff
@@ -0,0 +1,140 @@
+## The pincells - fuels first
+### Common primitives for defining the different fuel regions.
+fuel_pin_or = openmc.ZCylinder(r = geom.r_fuel)
+fuel_gap_1_or = openmc.ZCylinder(r = geom.r_fuel + geom.t_f_c_gap)
```
Is there a reason this is `fuel_gap_1_or` and not just `fuel_gap_or`?  I don't see any numbered 2, 3, etc?

nuclearkevin replied:
This is a carry over from the initial benchmark specification which had two layers of cladding on the fuel (stainless steel first, Zr second), which I believe was done to emulate the effects of grid spacers. I discarded the stainless steel cladding layer while simplifying the models (since it didn't add much when it comes to AMR), but neglected to change the naming scheme.

gonuke commented on models/lwr/openmc_pincells.py:
```diff
@@ -0,0 +1,140 @@
+mox_7_0_fuel_cell = openmc.Cell(name = '7.0% MOX Fuel Pin')
+mox_7_0_fuel_cell.region = -fuel_pin_or
+mox_7_0_fuel_cell.fill = mats.mox_7_0
+mox70_u = openmc.Universe(cells=[mox_7_0_fuel_cell, gap_1_cell_2, zr_clad_cell_2, h2o_bb_cell_2])
```
Showing my OpenMC ignorance: is it not possible to reuse the gap/clad/water cells in a different universe?

Or possible clone the cells from one use to the other?

nuclearkevin replied:
A cell can only be contained in a single universe (which is why these are duplicated). I didn't know that OpenMC had a clone function (definitely on me for not checking the cell documentation); I've replaced all of these with with `cells.clone()` to reduce code duplication, thanks for letting me know that it's an option!

gonuke commented on models/lwr/openmc_pincells.py:
```diff
@@ -0,0 +1,140 @@
+fuel_bb = openmc.model.RectangularPrism(width = geom.pitch, height = geom.pitch)
+
+### The entire 4.3% MOX pincell.
+gap_1_cell_1 = openmc.Cell(name = '4.3% MOX Pin Gap 1')
+gap_1_cell_1.region = +fuel_pin_or & -fuel_gap_1_or
+zr_clad_cell_1 = openmc.Cell(name = '4.3% MOX Pin Zr Clad')
+zr_clad_cell_1.region = +fuel_gap_1_or & -fuel_zr_or
+zr_clad_cell_1.fill = mats.zr
+h2o_bb_cell_1 = openmc.Cell(name = '4.3% MOX Pin Water Bounding Box')
+h2o_bb_cell_1.region = +fuel_zr_or & -fuel_bb
+h2o_bb_cell_1.fill = mats.h2o
+
+mox_4_3_fuel_cell = openmc.Cell(name = '4.3% MOX Fuel Pin')
+mox_4_3_fuel_cell.region = -fuel_pin_or
+mox_4_3_fuel_cell.fill = mats.mox_4_3
+mox43_u = openmc.Universe(cells=[mox_4_3_fuel_cell, gap_1_cell_1, zr_clad_cell_1, h2o_bb_cell_1])
```
I would recommend a more semantic data structure:
```
gap_cell['mox_4_3'] = ...
zr_clad_cell['mox_4_3'] = ...
etc...

fuel_cell['mox_4_3'] = ... 
```
(assuming you can't reuse cells that are the same across different universes - see question below)

gonuke commented on models/lwr/openmc_pincells.py:
```diff
@@ -0,0 +1,140 @@
+fuel_gap_1_or = openmc.ZCylinder(r = geom.r_fuel + geom.t_f_c_gap)
+fuel_zr_or = openmc.ZCylinder(r = geom.r_fuel + geom.t_f_c_gap + geom.t_zr_clad)
+fuel_bb = openmc.model.RectangularPrism(width = geom.pitch, height = geom.pitch)
+
+### The entire 4.3% MOX pincell.
+gap_1_cell_1 = openmc.Cell(name = '4.3% MOX Pin Gap 1')
+gap_1_cell_1.region = +fuel_pin_or & -fuel_gap_1_or
+zr_clad_cell_1 = openmc.Cell(name = '4.3% MOX Pin Zr Clad')
+zr_clad_cell_1.region = +fuel_gap_1_or & -fuel_zr_or
+zr_clad_cell_1.fill = mats.zr
+h2o_bb_cell_1 = openmc.Cell(name = '4.3% MOX Pin Water Bounding Box')
+h2o_bb_cell_1.region = +fuel_zr_or & -fuel_bb
+h2o_bb_cell_1.fill = mats.h2o
```
Can't you define these regions once and reuse:
```suggestion
gap_region = +fuel_pin_or & -fuel_gap_1_or
zr_clad_region = +fuel_gap_1_or & -fuel_zr_or
h2o_bb_region = +fuel_zr_or & -fuel_bb

### The entire 4.3% MOX pincell.
gap_1_cell_1 = openmc.Cell(name = '4.3% MOX Pin Gap 1', region=gap_region)
zr_clad_cell_1 = openmc.Cell(name = '4.3% MOX Pin Zr Clad', region=zr_clad_region, fill=mats.zr)
h2o_bb_cell_1 = openmc.Cell(name = '4.3% MOX Pin Water Bounding Box', region=h2o_bb_region, fill=mats.h2o)
```

And then similar lines for the subsequent similarly defined cells (assuming you can't just reuse or clone them)

gonuke reviewed (changes requested):
I like the reuse in the MOOSE files and some of the python... a few more suggestions here, just focussing on the pincells for now

nuclearkevin commented on models/lwr/pincells.i:
```diff
@@ -0,0 +1,144 @@
+    ring_intervals = '${FUEL_RADIAL_DIVISIONS} 1 1'
+    polygon_size = ${fparse pitch / 2.0}
+
+    ring_block_ids = '4 5 2 3'
+    ring_block_names = 'mox_43_center mox_43 gap_1 zr_clad'
```
Unfortunately we can't. The `PolygonConcentricCircleMeshGenerator` contains specific metadata which is required by `PatternedCartesianMeshGenerator` to generate assembly/core meshes. This metadata gets deleted if the mesh gets passed to a `RenameBlockGenerator`, and so `PatternedCartesianMeshGenerator` throws an error.

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,65 @@
+for layer_idx, planes in enumerate(zip(core_z_planes[:-1], core_z_planes[1:])):
+  layer = +planes[0] & -planes[1]
+  all_cells.append(openmc.Cell(fill = mats['UO2'],  region=-pins.fuel_pin_or   & layer,                       name=f'Fuel {layer_idx}'))
+  all_cells.append(openmc.Cell(fill = None,         region=+pins.fuel_pin_or   & -pins.fuel_gap_1_or & layer, name=f'Gap {layer_idx}'))
+  all_cells.append(openmc.Cell(fill = mats['ZR_C'], region=+pins.fuel_gap_1_or & -pins.fuel_zr_or & layer,    name=f'Clad {layer_idx}'))
```
Also a weak recommendation to define regions in the pincell file:
```
fuel_region = -fuel_pin_or
gap_region = +fuel_pin_or & -fuel_gap_or
#etc...
```

And then use these everywhere for consistency.  It's probably not a huge benefit, but does reduce some of the cognitive burden in reading/reviewing/error checking

gonuke commented on models/lwr/openmc_assemblies.py:
```diff
@@ -0,0 +1,148 @@
+uo2_assembly = openmc.RectLattice(name = 'UO2 Assembly')
+uo2_assembly.pitch = (geom.pitch, geom.pitch)
+uo2_assembly.lower_left = (-17.0 * geom.pitch / 2.0, -17.0 * geom.pitch / 2.0)
+uo2_assembly.universes = [
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 1
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 2
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 3
+  [p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2']], # 4
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 5
+  [p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2']], # 6
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 7
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 8
+  [p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['FIS'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2']], # 9
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 10
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 11
+  [p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2']], # 12
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 13
+  [p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2']], # 14
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['GTB'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 15
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']], # 16
+  [p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2'], p['UO2']]  # 17
+]# 1         2         3         4         5         6         7         8         9         10        11        12        13        14        15        16        17
```
A weak suggestion that these could just be lists of strings to make it slightly less verbose, followed by:
```
uo2_assembly.universes = [ [ p[pin_type] for pin_type in row] for row in assembly_map ]
```
It doesn't buy a lot (hence making is a "weal" suggestion), but makes things a little more compact and readable?

nuclearkevin replied:
My preference would be to keep the assembly layout this way as it mirrors the setup for the mesh mirrors in `assemblies.i` and how the assembly layouts are defined in the original benchmark.

gonuke replied:
To be more clear in case there was confusion about my suggestion... There would still be this layout, but instead of it being objects like `p['UO2']`, the entries would just be the strings `'UO2'`.  It's slightly more compact and readable.

The extra line just converts these strings into their pincell objects....

nuclearkevin replied:
Ah, I see what you mean now. Sorry about the misunderstanding, I'll get that implemented shortly!

gonuke commented on models/lwr/3x3_lattice/openmc.i:
```diff
@@ -0,0 +1,53 @@
+  batches = 10000
+
+  verbose = true
+  power = ${fparse 3000e6 / 273 / (17 * 17) * 9}
```
I see this line in other files - could it be defined in `common.i`??

nuclearkevin replied:
The pin power (`3000e6 / 273 / (17 * 17)`) could be defined in `common.i` and then used in each of the `openmc.i` files to define `power`. Actual value of `power` in each case is different.

nuclearkevin replied:
Thinking about it some more, I would prefer to avoid placing parameters from `openmc.i` in `common.i` as most of those parameters are unused. MOOSE throws warnings to the console (see below) which may be alarming to people who're running the cases.
```
*** Warning ***
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:25.1: unused parameter 'pitch'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:28.1: unused parameter 'core_height'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:31.1: unused parameter 'reflector_t'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:34.1: unused parameter 'r_fuel'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:37.1: unused parameter 't_f_c_gap'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:40.1: unused parameter 't_zr_clad'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:43.1: unused parameter 'r_guide'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:46.1: unused parameter 't_al_clad'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:52.1: unused parameter 'NUM_SECTORS'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:53.1: unused parameter 'FUEL_RADIAL_DIVISIONS'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:54.1: unused parameter 'BACKGROUND_DIVISIONS'
/home/kevin/Projects/cardinal_amr/models/lwr/multi_assembly/unrodded/../../common.i:55.1: unused parameter 'AXIAL_DIVISIONS'
```

gonuke commented on models/lwr/openmc_assemblies.py:
```diff
@@ -0,0 +1,148 @@
+### UO2 fueled assembly.
+uo2_assembly = openmc.RectLattice(name = 'UO2 Assembly')
+uo2_assembly.pitch = (geom.pitch, geom.pitch)
+uo2_assembly.lower_left = (-17.0 * geom.pitch / 2.0, -17.0 * geom.pitch / 2.0)
```
Magic number 17 - define this as a variable?

gonuke commented on models/lwr/3x3_lattice/make_openmc_model.py:
```diff
@@ -0,0 +1,68 @@
+
+assembly = openmc.RectLattice(name='Fuel Assembly')
+assembly.pitch = (geom.pitch, geom.pitch)
+assembly.lower_left = (-3.0 * geom.pitch / 2.0, -3.0 * geom.pitch / 2.0)
```
Magic number 3: define this as a variable?

gonuke commented on models/lwr/3x3_lattice/make_openmc_model.py:
```diff
@@ -0,0 +1,68 @@
+assembly.pitch = (geom.pitch, geom.pitch)
+assembly.lower_left = (-3.0 * geom.pitch / 2.0, -3.0 * geom.pitch / 2.0)
+assembly.universes = [
+  [pins.uo2_u, pins.uo2_u, pins.uo2_u],
```
This seems to need to be updated to new syntax?

gonuke commented on models/lwr/17x17_uo2_assembly/rodded/mesh.i:
```diff
@@ -0,0 +1,24 @@
+[Mesh]
+  [Delete_Blocks]
+    type = BlockDeletionGenerator
+    input = UO2_Assembly_Rodded
```
Is this the only line that is different between this file and the unrodded equivalent? Could it be handled with some kind of reuse? (Not sure how)

nuclearkevin replied:
I don't think there's a way for us to avoid code duplication for the rodded/unrodded 17x17 assemblies here unless we want to extrude all of the 2D assemblies into 3D meshes before deleting the blocks, which just shifts it into `assemblies.i`.

gonuke replied:
I was just curious about whether an `fparse` can resolve to the name of an input so that this could be:
```suggestion
    input = ${fparse assembly_type}
```

Perhaps not...

gonuke reviewed (changes requested):
another round of streamlining/reuse...

I wonder if it makes sense to have a `case.i` with problem specific values that can be parsed into some  common files to be reused?

aprilnovak commented on models/lwr/17x17_uo2_assembly/rodded/make_openmc_model.py:
```diff
@@ -0,0 +1,57 @@
+from openmc_assemblies import assembly_bb, pins_per_axis
+
+ap = ArgumentParser()
+ap.add_argument('-n', dest='n_axial', type=int, default=1,
```
Can we add something to the README.md that describes this flag? We're not really exercising it now, but it may be useful to have an explicit note about it.

nuclearkevin commented:
I would prefer to keep each OpenMC input file separate with minimal includes. This way users can modify the Cardinal OpenMC run without needing to worry about the propagation of changes from one input file to another.

gonuke commented:
> I would prefer to keep each OpenMC input file separate with minimal includes. This way users can modify the Cardinal OpenMC run without needing to worry about the propagation of changes from one input file to another (which maximizes flexibility when testing AMR strategies).

Do you mean the OpenMC python or the OpenMC `.i` files for Cardinal/Moose? or both?

gonuke commented on models/lwr/multi_assembly/rodded/openmc.i:
Could this be defined once for both the rodded and unrodded and then just symlinked into the relevant directory? Similarly for each case that includes rodded/unrodded....

nuclearkevin replied:
See my comment [here](https://github.com/neams-th-coe/cardinal_amr/pull/1#issuecomment-2679862805).

gonuke reviewed (changes requested):
A few final (?) suggestions....  I still see some opportunities for code reuse, but you have reasons for keeping things distinct.

nuclearkevin commented:
> Do you mean the OpenMC python or the OpenMC .i files for Cardinal/Moose? or both?

Just the `openmc.i` files. I envision that users will want to be able to tune the parameters of the OpenMC solve on a per-problem basis (at least that's what I've been doing with my AMR experiments), and so keeping a distinct `openmc.i` file with minimal includes (other than the AMR strategy) would be preferable for these.

gonuke commented on models/lwr/pincell/make_openmc_model.py:
```diff
@@ -0,0 +1,65 @@
+  all_cells.append(openmc.Cell(fill = mats['UO2'],  region = pins.fuel_pin_region & layer,                     name=f'Fuel {layer_idx}'))
+  all_cells.append(openmc.Cell(fill = None,         region = pins.fuel_gap_region & -pins.fuel_gap_or & layer, name=f'Gap {layer_idx}'))
+  all_cells.append(openmc.Cell(fill = mats['ZR_C'], region = pins.fuel_clad_region & -pins.fuel_zr_or & layer, name=f'Clad {layer_idx}'))
+  all_cells.append(openmc.Cell(fill = mats['H2O'],  region = pins.fuel_water_region & layer & -pins.fuel_bb,   name=f'Water {layer_idx}'))
```
I think `pins.fuel_water_region` already includes `-pins.fuel_bb`?
```suggestion
  all_cells.append(openmc.Cell(fill = mats['H2O'],  region = pins.fuel_water_region & layer ,   name=f'Water {layer_idx}'))
```

nuclearkevin replied:
Good catch, thanks!

gonuke reviewed (changes requested):
This all looks great given some of your specific design choices.  I found one last tweak and then I'll be ready to approve

gonuke reviewed (approved):
Thanks for working through all the streamlining @nuclearkevin - this looks good now

nuclearkevin commented:
Thanks for the thorough review @gonuke! 

@aprilnovak this should be good to go now unless you have additional comments / changes you want implemented.
